### PR TITLE
Eliom ppx

### PR DESCRIPTION
**Pull request by Drup**
It's starting to get into a decent shape, so here it is. Documentation is included.

I used the paper for documentation, instead of the camlp4 documentation, since I think it's much simpler and straight to the point. Maybe I will backport that.

Two remaining points
- test on big examples (I only did small things)
- Figure out whether we need the nested_* stuff.

@hhugo: please review the build/packaging part.

### CODE REVIEW

**Comment by Drup** [comment]
[Link to the documentation](https://github.com/ocsigen/eliom/blob/ppx/src/ppx/ppx_eliom.mli)

File: pkg/filelist.ml [hunk 1]
```diff
@@ -139,6 +139,12 @@ let ocamlbuild = {
 
 }
 
+let ppx = {
+  interface = [ "ppx_eliom" ; "ppx_eliom_client" ; "ppx_eliom_type" ; "ppx_eliom_server" ];
+  internal = [ "ppx_eliom_utils" ];
+}
```
Comment (vasilisp) [inline pkg/filelist.ml]: Add an `interface_only` field here. This is for `mli`-only modules, which have started appearing in Eliom. This fixes the build failure

File: src/lib/eliom_pervasives.client.ml [hunk 2]
```diff
@@ -1,6 +1,8 @@
 
 type 'a client_value = 'a Eliom_lib.client_value
 
+type 'a fragment = 'a client_value
```
Comment (vasilisp) [inline src/lib/eliom_pervasives.client.ml]: Do we really need a new name? "Fragment" could just as well indicate an `str_item`.

**Comment by vasilisp** [comment]
I don't think the `nested_*` stuff is needed, since you expand out shared expressions. At least for now; if we allow for arbitrary nesting as per #196, the need for separation may re-arise.

**Comment by Drup** [comment]
I'm pretty sure the nesting already works, actually.

File: src/lib/eliom_pervasives.client.ml (hunk 2 above)
Comment (Drup) [inline src/lib/eliom_pervasives.client.ml]: I intend to progressively change the name everywhere over the next versions.

**Comment by vasilisp** [comment]
`js_of_eliom` tries to pipe the shared file `eba_group.mli` through the client preprocessor. As a result, the interface ends up almost empty and EBA fails to compile.

**Comment by Drup** [comment]
@vasilisp I'm wondering how that could work with camlp4. It's basically the same process.

**Comment by vasilisp** [comment]
@Drup, the process was subtly different :). 832bc0dec3bd6286fc606080c5bb283df360460f fixes the issue for me.

**Comment by Drup** [comment]
I see, I still don't really get why it works in camlp4.

**Comment by vasilisp** [comment]
I can't figure out how to use `eliomc` et al. on code that uses other PPX extensions. Preprocessing happens via the `ppx_eliom_*` binaries which (as far as I can tell) do not allow loading other extensions.

**Comment by vasilisp** [comment]
You can find a few more patches in the `ppx-fix-eliomc` branch (for now it is quick fixes, I will clean things up later).

Also, there is an issue with type inference. Given injections `~%bar` and `~%Foo.bar`, `Foo.bar` gets the same type as `bar`.

**Comment by Drup** [comment]
The name collision issue should be fixed.

**Comment by vasilisp** [comment]
With some `sed` preprocessing (coupled with b3b5561609c162b344a361151b6c9217be09d567) to deal with non-parsable `type_mli` files, and with a hackish replacement for `eliomdep -sort` (needed to handle a mix of Camlp4 and PPX files), I have managed to get our mid-sized Camlp4-to-PPX "case study" to compile. Now there are issues with missing injections and closures on the client, and I have no clue what is going on.

**Comment by Drup** [comment]
Can you show me the piece of code involved ?

**Comment by Drup** [comment]
Apparently, it's all fixed now.

File: src/ppx/ppx_eliom_utils.ml [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+        dispatch_str !context mapper pstr
+    in
+    let loc = {(Location.in_file !Location.input_name) with loc_ghost = true} in
+    open_eliom_pervasives :: Pass.prelude loc @ flatmap f structs @ Pass.postlude loc
+
+  let toplevel_signature context mapper sigs =
+    let f psig =
+      let loc = psig.psig_loc in
+      match psig.psig_desc with
+      | Psig_extension (({txt=("shared"|"client"|"server" as txt)}, PStr strs), _) ->
```
Comment (vasilisp) [inline src/ppx/ppx_eliom_utils.ml]: @Drup, this should be `"shared.start"|"client.start"|"server.start"` (consistently with structures and the documentation), right?

**Comment by vasilisp** [comment]
As a further test, I converted `ocsigen-toolkit` to PPX: ocsigen/ocsigen-toolkit@53b141961eb1008a74a550b4a63665e814aa5e27 . The conversion was trouble-free, modulo small glitches with the `.eliomi` files (which are easy to fix).

File: src/ppx/ppx_eliom_utils.ml (hunk 3 above)
Comment (Drup) [inline src/ppx/ppx_eliom_utils.ml]: Indeed.

Comment (Drup) [inline src/ppx/ppx_eliom_utils.ml]: Fixed.

**Comment by Drup** [comment]
@vasilisp Great. I think we are pretty much ready to merge.

**Comment by vasilisp** [comment]
@Drup, can we merge? If any issues arise, we can fix them in master.

My mind cannot handle too much pinning and too many switches :).

**Comment by Drup** [comment]
Absolutely. I was expecting you to do when done with testing. :)

**Pull request merged**